nemo/collections/nlp/models/language_modeling/megatron_retrieval_model.py

```diff
@@ -0,0 +1,521 @@
+        retrieved_ids = batch['retrieved_ids']
+        retrieved_attn_mask = batch['retrieved_emb_mask']
+        labels = batch['labels']
+        loss = self(input_tokens_id, input_attn_mask, retrieved_ids, retrieved_attn_mask, labels=labels)
+        loss_mask = loss_mask.float()
+        lm_loss = torch.sum(loss.view(-1) * loss_mask.reshape(-1)) / loss_mask.sum()
+        reduced_loss = average_losses_across_data_parallel_group([lm_loss])
+        return reduced_loss
+
+    def validation_epoch_end(self, outputs):
+        if not outputs:
+            return
+        averaged_loss = torch.stack(outputs).mean()
+        self.log('val_loss', averaged_loss, prog_bar=True)
+        self.log('consumed_samples', self.compute_consumed_samples(self.trainer.global_step - self.init_global_step))
+        return averaged_loss
+
+    def test_step(self, batch, batch_idx):
+        return self.validation_step(batch, batch_idx)
+
+    def test_epoch_end(self, outputs):
+        averaged_loss = average_losses_across_data_parallel_group(outputs)
+        logging.info(f'test_loss: {averaged_loss[0]}')
+        self.log(
+            'consumed_samples', self.compute_consumed_samples(self.trainer.global_step - self.init_global_step),
+        )
+        return averaged_loss
+
+    def build_train_valid_test_datasets(self):
+        logging.info('Building RETRO datasets.')
+        self._train_ds, self._validation_ds, self._test_ds = build_mock_train_valid_test_datasets(
+            cfg=self.cfg,
+            trainer=self.trainer,
+            splits_string=self.cfg.data.splits_string,
+            tokenizer=self.tokenizer,
+            mock_data_size=self.cfg.data.get('mock_data_size', 10000),
+        )
+        if self._train_ds is not None:
+            logging.info(f'Length of train dataset: {len(self._train_ds)}')
+        if self._validation_ds is not None:
+            logging.info(f'Length of val dataset: {len(self._validation_ds)}')
+        if self._test_ds is not None:
+            logging.info(f'Length of test dataset: {len(self._test_ds)}')
+        logging.info(f'Finished building RETRO datasets.')
+        return self._train_ds, self._validation_ds, self._test_ds
+
+    def build_pretraining_data_loader(self, dataset, consumed_samples):
+        """Buld dataloader given an input dataset."""
+
+        if dataset is None:
+            return None
+
+        logging.info(f'Building dataloader with consumed samples: {consumed_samples}')
+        # Megatron sampler
+        if hasattr(self._cfg.data, 'dataloader_type') and self._cfg.data.dataloader_type is not None:
+            if self._cfg.data.dataloader_type == 'single':
+                batch_sampler = MegatronPretrainingBatchSampler(
+                    total_samples=len(dataset),
+                    consumed_samples=consumed_samples,
+                    micro_batch_size=self._cfg.micro_batch_size,
+                    global_batch_size=self._cfg.global_batch_size,
```

Comment:
If not using pipeline parallel, then you want to check explicitly to make sure that global batch is always equal to micro batch * data_parallel_world_size. Do you check for this anywhere?

Reply:
use the `MegatronPretrainingSampler` sampler